### PR TITLE
Auth/Sessions/Routing

### PR DESCRIPTION
**Pull request by pharapeti**
## What is this change doing?

- Fixes model associations
- Sets up routes, see `server.js` - i've split up the routes into /auth, /admin and /student
- Using encrypted password instead of plaintext for authentication via `bcryptjs`
- Persistent session store, allows a user to stay logged in after server restart

## Why is it needed?

In same order as above

- So we can use the associations to be able to make complex queries
- Allows us to break up the authentication, admin and student related endpoints - less complexity
- So we aren't storing plaintext passwords in the db
- Acts more like a real web app

To authenticate users + routing for future endpoints

## Where should the reviewer start?

server.js

## How can these changes be manually tested?

1) Use postman and check that (most 😉 ) endpoints are returning the right data
2) Ensure you stay logged in even after restarting the server
3) Ensure no plaintext passwords are being stored in the db after logging in

## What user story does this PR relate to? (include link)

Bunch of user stories

### CODE REVIEW

File: controllers/admin/groups/groups.controller.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    if (group == null) {
+      res.status(404).send({ message: `Group with id: ${req.params.id} does not exist` });
+    } else {
+      res.send(group);
+    }
+  })
+  .catch(err => {
+    res.status(500).send({ message: err.message || 'Something went wrong' });
+  })
+}
```
Comment (DaGev) [inline controllers/admin/groups/groups.controller.js:30]: Does indeed return all groups. Could it be a good idea to allow an admin to be shown the groups belonging to his projects differentiated in some way?

File: controllers/admin/projects/projects.controller.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    if (project == null) {
+      res.status(404).send({ message: `Project with id: ${req.params.id} does not exist` });
+    } else {
+      res.send(project);
+    }
+  })
+  .catch(err => {
+    res.status(500).send({ message: err.message || 'Something went wrong' });
+  })
+}
```
Comment (DaGev) [inline controllers/admin/projects/projects.controller.js:30]: Does return the projects but again, shouldn't his own projects maybe be highlighted?

File: controllers/student/profile/profile.controller.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    include: [
+      { model: model.GroupAllocation },
+      { model: model.PreferenceSelection }
+    ]
+   }).then(user => {
+     res.send(user);
+   }).catch(err => {
+     res.status(500).send({ message: err.message || 'Something went wrong' })
+   })
+}
```
Comment (DaGev) [inline controllers/student/profile/profile.controller.js:16]: Works, but also works with admins, not just students. Need to be changed or remain?

File: controllers/student/groups/groups.controller.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    if (group == null) {
+      res.status(404).send({ message: `Group with id: ${req.params.id} does not exist` });
+    } else {
+      res.send(group);
+    }
+  })
+  .catch(err => {
+    res.status(500).send({ message: err.message || 'Something went wrong' });
+  })
+}
```
Comment (DaGev) [inline controllers/student/groups/groups.controller.js:53]: Returns all groups, not just the ones the student is in, BUT in controllers/student/groups/profile/profile.controller.js , the displaying of GroupAllocation model works fine.

Student is able to get access group information of groups he's not allocated to.

File: controllers/student/projects/projects.controller.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    })
+    .catch(err => {
+      res.status(500).send({ message: err.message || "Some error occurred while retrieving projects." });
+    });
+}
+
+// Return information about a single project
+exports.findOne = (req, res) => {
+  res.send([]);
+}
```
Comment (DaGev) [inline controllers/student/projects/projects.controller.js:61]: Doesn't display project information individual or list.

**Review by DaGev — CHANGES_REQUESTED** [review]
Not yet ready to be merged, I left comments on some of the files indicating what the issues are.

File: controllers/admin/projects/projects.controller.js (hunk 2 above)
Comment (pharapeti) [reply controllers/admin/projects/projects.controller.js:30]: I'd say it's implied based on the route the admin will be making a request to. Don't think we can really add much to differentiate the difference

File: controllers/student/projects/projects.controller.js (hunk 5 above)
Comment (pharapeti) [reply controllers/student/projects/projects.controller.js:61]: Done

File: controllers/admin/groups/groups.controller.js (hunk 1 above)
Comment (pharapeti) [reply controllers/admin/groups/groups.controller.js:30]: Yep we should implement that, can be done in another ticket

File: controllers/student/profile/profile.controller.js (hunk 3 above)
Comment (pharapeti) [reply controllers/student/profile/profile.controller.js:16]: that's fine as it is, we can always update if we wish to change the logic

**Comment by pharapeti** [comment]
@DaGev feedback addressed, please have a look then approve if ready

File: controllers/student/groups/groups.controller.js (hunk 4 above)
Comment (pharapeti) [reply controllers/student/groups/groups.controller.js:53]: Done!

**Review by DaGev — APPROVED** [review]

**Pull request merged**